### PR TITLE
Added suite tests for the reallocator and fixed nil map edge cases

### PR DESCRIPTION
**Pull request by njtran**
Issue #, if available:

Description of changes:

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+			ExpectCreated(ns.Client, p)
+			ExpectEventuallyReconciled(ns.Client, p)
+
+			Expect(ns.Client.Get(context.Background(), client.ObjectKey{Name: node.Name}, node)).To(Succeed())
+			Expect(node.Labels).To(HaveKeyWithValue(v1alpha1.ProvisionerPhaseLabel, v1alpha1.ProvisionerUnderutilizedPhase))
+			Expect(node.Annotations).To(HaveKey(v1alpha1.ProvisionerTTLKey))
+		})
+
+		It("should remove labels from utilized nodes", func() {
```
Comment (JacobGabrielson) [inline pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go:97]: I think it would be good to add a test that cycles nodes thru the states a few times? Like say it starts out underutilized, then goes to utilized, then goes back to underutilized?

Comment (njtran) [reply pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go:102]: Good idea. I think an e2e test would be good too.

File: pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest/printer"
+)
+
+func TestAPIs(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecsWithDefaultAndCustomReporters(t,
+		"Provisioner",
```
Comment (ellistarn) [inline pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go:43]: Fixme

File: pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+		ns, err = env.NewNamespace()
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		ExpectCleanedUp(ns.Client)
+	})
+
+	Context("Reconciliation", func() {
+		It("should label nodes for underutilized", func() {
```
Comment (ellistarn) [inline pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go:79]: "label for underutilized" is really wonky english. "label as"?

File: pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+			p := &v1alpha1.Provisioner{
```
Comment (ellistarn) [inline pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go:80]: nit: name this provisioner

File: pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+				Conditions: []v1.PodCondition{{Type: v1.PodReady, Status: v1.ConditionTrue}},
+			})
+
+			ExpectCreatedWithStatus(ns.Client, node)
+			ExpectCreatedWithStatus(ns.Client, pod)
+
+			ExpectCreated(ns.Client, p)
+			ExpectEventuallyReconciled(ns.Client, p)
+
+			nodeCheck := &v1.Node{}
```
Comment (ellistarn) [inline pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go:132]: I think you can reuse the node object if you wanted to. Otherwise I might name this updatedNode

File: pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+					v1alpha1.ProvisionerTTLKey: time.Now().Add(time.Duration(-100) * time.Second).Format(time.RFC3339),
+				},
+			})
+			ExpectCreatedWithStatus(ns.Client, node)
+
+			ExpectCreated(ns.Client, p)
+			ExpectEventuallyReconciled(ns.Client, p)
+
+			nodeCheck := &v1.Node{}
+			Expect(ns.Client.Get(context.Background(), client.ObjectKey{Name: node.Name}, nodeCheck)).ToNot(Succeed())
```
Comment (ellistarn) [inline pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go:161]: This may fail in a spurious way (like a network connectivity issue). I wonder if we could write a check for does not exist? expectations.ExpectToNotExist(client, object). There's a good example of the NotFound error checking in the genericcontroller

File: pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
```
Comment (ellistarn) [inline pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go:187]: same for not exists here

File: pkg/test/objects.go [hunk 8]
```diff
@@ -60,28 +61,35 @@ func PodWith(options PodOptions) *v1.Pod {
 	if options.Image == "" {
 		options.Image = "k8s.gcr.io/pause"
 	}
+
+	spec := &v1.PodSpec{
```
Comment (ellistarn) [inline pkg/test/objects.go:65]: Keep this inline

File: pkg/test/objects.go [hunk 9]
```diff
@@ -60,28 +61,35 @@ func PodWith(options PodOptions) *v1.Pod {
...
+			Name:  options.Name,
+			Image: options.Image,
+			Resources: v1.ResourceRequirements{
+				Requests: options.ResourceRequests,
+			},
+		}},
+	}
+
+	if options.NodeName != nil {
+		spec.NodeName = *options.NodeName
```
Comment (ellistarn) [inline pkg/test/objects.go:77]: since nodename is a string, if unspecified, it'll be "", so you can inline it.

File: pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go (hunk 1 above)
Comment (njtran) [reply pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go:102]: Punting on this for a follow-up PR.

File: pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go (hunk 6 above)
Comment (njtran) [reply pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go:161]: I will use this. Thanks.

File: pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go (hunk 5 above)
Comment (njtran) [reply pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go:132]: Since this test requires deleting the labels and annotations, I think there's some weird merge-like behavior happening in the Get function if I reuse node here. Every other test works fine without this, but I'll change this usage to be `updatedNode`

**Review by JacobGabrielson — APPROVED** [review]

**Comment by JacobGabrielson** [comment]
```
    __|__ |___| |\
    |o__| |___| | \
    |___| |___| |o \
   _|___| |___| |__o\
  /...\_____|___|____\_/
  \   o * o * * o o  /
~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: pkg/controllers/provisioning/v1alpha1/allocation/suite_test.go [hunk 10]
```diff
@@ -39,7 +39,7 @@ import (
 func TestAPIs(t *testing.T) {
 	RegisterFailHandler(Fail)
 	RunSpecsWithDefaultAndCustomReporters(t,
-		"Provisioner",
+		"Provisioner-Allocator",
```
Comment (ellistarn) [inline pkg/controllers/provisioning/v1alpha1/allocation/suite_test.go:42]: Hmm, WDYT about Provisioner/Allocation? In general, I like the idea of parity between package name and suite name. After all, this is the suite for this package.

File: pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+		var err error
+		ns, err = env.NewNamespace()
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		ExpectCleanedUp(ns.Client)
+	})
+
+	Context("Reconciliation", func() {
```
Comment (ellistarn) [inline pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go:79]: Nit: Think about your contexts and your describe blocks. Is there redundancy here? What's a sane model for naming all of these things?

**Review by ellistarn — APPROVED** [review]

File: pkg/controllers/provisioning/v1alpha1/allocation/suite_test.go (hunk 10 above)
Comment (njtran) [reply pkg/controllers/provisioning/v1alpha1/allocation/suite_test.go:42]: I like the /, it keeps in line with other distinctions in annotation that we've done. I'll make that change.

File: pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go (hunk 11 above)
Comment (ellistarn) [reply pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go:79]: misread and thought Reconciliation Reallocation were the same word. Sorry! Nice work.

**Review by ellistarn — APPROVED** [review]
